src/lib/yang/path_data.lua

```diff
@@ -524,6 +528,64 @@ function consistency_checker_from_grammar(grammar)
                ("Broken leafref integrity in '%s' when referencing '%s'"):format(
                 path, leafref))
       end
+   end   
+end
+
+function uniqueness_checker_from_grammar(grammar)
+   -- Generate checker for table
+   local function unique_assertion(leaves, grammar)
+      local unique_leaves = {}
+      for leaf in leaves:split(" +") do
+         table.insert(unique_leaves, normalize_id(leaf))
+      end
+      local pairs = grammar.key_ctype and cltable.pairs or pairs
+      return function (tab)
+         -- Sad quadratic loop, again
+         for k1, v1 in pairs(tab) do
+            for k2, v2 in pairs(tab) do
+               if k1 == k2 then break end
+               local collision = true
+               for _, leaf in ipairs(unique_leaves) do
+                  if not lib.equal(v1[leaf], v2[leaf]) then
+                     collision = false
+                     break
+                  end
+               end
+               assert(not collision, "Not unique: "..leaves)
+            end
+         end
```

Comment:
I'm not sure about this.

It seems that the leaf unique check conditions checks whether each of the unique fields has an unique value, however the multiple unique values are provide the uniqueness applies to the combination of all the unique fields.

```
The "unique" constraint specifies that the combined values of all the
   leaf instances specified in the argument string, including leafs with
   default values, MUST be unique within all list entry instances in
   which all referenced leafs exist or have default values.  The
   constraint is enforced according to the rules in Section 8.
```

https://tools.ietf.org/html/rfc7950

In addition, I think it could be possible to get rid of the quadratic loop by building a reverse hash by value map as the table is iterated. I coded an example:

```lua
local t1 = {
   {ip = "192.168.0.1", port = "80"},
   {ip = "192.168.0.2", port = "80"},
}
local t2 = {
   {ip = "192.168.0.1", port = "80"},
   {ip = "192.168.0.1", port = "80"},
}

local unique_keys = {'ip', 'port'}

local function check (t)
   local unique_values = {}
   for _, each in ipairs(t) do
      p = unique_values
      for i, key in ipairs(unique_keys) do
         local value = assert(each[key])
         if i == #unique_keys then
            if p[value] then
               error("Value is not unique: "..value)
            end
            p[value] = true
         else
            if not p[value] then
               p[value] = {} 
            end
            p = p[value]
         end
      end
   end
end
  
check(t1)
print("t1 ok")
check(t2)
print("t2 ok")
```

Leaving this aside, the rest of the PR looks good to me.

Reply:
The way I parse the RFC:

- `unique "foo bar baz";` means every instance must have a unique combination of `foo`, `bar`, and `baz`. I.e.: `list {foo 1; bar 2; baz 3 } list {foo 1; bar 2; baz 42;}` is valid.
- `unique "foo"; unique "baz;" means every instance must have a unique leaves `foo` and `bar`. I.e.: `list {foo 1; baz 2;} list {foo 3; baz 4;}` is valid.

The code intends to iterate over all entry pairs and tries to find a difference in the unique leaves, if there is none a collision is found. I am afraid I don’t really follow your hash table scheme, but shouldn’t it fail on `t2`?

Reply:
Yes, the example fails on t2 ("t2 ok" is not printed since the non-unique error is triggered).

Right, I didn't know an `unique` statement could occur more than once within a `list` element, although that's possible according to the YANG spec (https://tools.ietf.org/html/rfc7950) (cardinality of unique is 0..n).

Basically the code snippet i posted what it does is to build a map of unique elements as a list is traversed. When attempting to insert a new element that already exists in the map an error is triggered (the element is duplicated). That saves the n^2 loop as the list is only traversed once (although there's a nested loop for the list of unique elements).

I changed the code to make it work with a list of possible unique elements (covers 2nd bullet):

```lua
local t1 = {
   {ip = "192.168.0.1", port = "80"},
   {ip = "192.168.0.2", port = "80"},
}
local t2 = {
   {ip = "192.168.0.1", port = "80"},
   {ip = "192.168.0.1", port = "80"},
}

local unique_key_list = {
   {'ip'},
   {'port'},
   {'ip', 'port'},
}
    
local function check (t)
   for _, unique_key in ipairs(unique_key_list) do
      local unique_values = {}
      for _, each in ipairs(t) do
         local p = unique_values
         for i, key in ipairs(unique_key) do
            local value = key.."="..each[key]
            if i == #unique_key then
               if p[value] then
                  error("Value is not unique: "..value)
               end
               p[value] = true
            else
               if not p[value] then
                  p[value] = {} 
               end
               p = p[value]
            end
         end
      end
   end
end
  
check(t1)
print("t1 ok")
check(t2)
print("t2 ok")
```

So for t1, when checking `unique=ip` the list would look like:

```lua
unique_values = {
   'ip=192.168.0.1' = true,
   'ip=192.168.0.2' = true,
}
```

When checking `unique=port' an error is detected:

```lua
unique_values = {
   'port=80' = true,
}
```
When attempting to insert 'port=80' again (second row), an error is triggered.

A list with composed unique fields ({ip, port}) would look like this:

```lua
unique_values = {
   'ip=192.168.0.1' = {
      'port=80' = true,
   }
   'ip=192.168.0.2' = {
      'port=80' = true,
   }
}
```

Reply:
Right, I see. But: the original code uses `lib.equal` so it works on cdata which I don’t think your version can handle (IIRC cdata doesn’t work for Lua table keys).

I also don’t know if its worth the complexity to optimize this at all? For my use cases its fine as is, but I can definitely see it becoming a problem if you have a huge set of lwaftr bindings modeled in a list with unique restrictions. If so, I feel like this can be done in a separate PR when the problem arises.

Just in case this got lost: the original code behaves the same AFAICT, it just doesn’t cache its work. ;-) I.e., it starts off with `collision = true` and any difference in the unique members will set it to `false`.

Reply:
Sorry for the delay dealing with this.

I'm fine with landing your patch and reworking the code that validates for uniqueness in another PR, as suggested.

OTOH, snabb-bot reports an error while trying to merge this PR in the following test: `program/lwaftr/tests/selftest.sh`.

I run the test manually to check where might be the issue:

```bash
$ sudo SNABB_PCI0=0000:82:00.0 SNABB_PCI1=0000:02:00.0 program/lwaftr/tests/selftest.sh 
...
test_add (subcommands.config_test.TestConfigMisc) ... FAIL
...
======================================================================
FAIL: test_add (subcommands.config_test.TestConfigMisc)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dpino/workspace/snabb_host/src/program/lwaftr/tests/subcommands/config_test.py", line 412, in test_add
    self.run_cmd(add_args)
  File "/home/dpino/workspace/snabb_host/src/program/lwaftr/tests/test_env.py", line 96, in run_cmd
    self.fail('\n'.join(msg_lines))
AssertionError: Error running command:
/home/dpino/workspace/snabb_host/src/snabb config add --schema=snabb-softwire-v2 config-test-daemon /softwire-config/binding-table/softwire { ipv4 8.8.8.8; psid 7; b4-ipv6 ::2; br-address 2001:db8::;port-set { psid-length 16; }}
Daemon Command:
/home/dpino/workspace/snabb_host/src/snabb lwaftr bench --bench-file /dev/null --name config-test-daemon /home/dpino/workspace/snabb_host/src/program/lwaftr/tests/data/icmp_on_fail.conf /home/dpino/workspace/snabb_host/src/program/lwaftr/tests/benchdata/ipv4-0550.pcap /home/dpino/workspace/snabb_host/src/program/lwaftr/tests/benchdata/ipv6-0550.pcap
Exit code: 1
STDOUT
lib/yang/path_data.lua:566: attempt to index local 'datum' (a number value)

STDERR
```

The test works fine in `lwaftr` branch, so there might be something in this PR that is affecting that test (perhaps the test needs to be changed or the PR code).

Reply:
I have pushed a commit that should fix the test failure (the culprit was broken handling of c(l)tables in `uniqueness_checker_from_grammar`.)